### PR TITLE
Fix for UTF-8 BOM

### PR DESCRIPTION
**Pull request by anirudh2290**
Ignore UTF-8 BOM character in CSV Parser if present.

@tqchen @piiswrong

### CODE REVIEW

**Comment by anirudh2290** [comment]
@cjolivier01 @reminisce @rahul003 @eric-haibin-lin  Can you guys please help review ?

File: src/data/text_parser.h [hunk 1]
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
...
+    * \return true if the data is loaded, false if reach end
+    */
+  inline bool FillData(std::vector<RowBlockContainer<IndexType>> *data);
+   /*!
+    * \brief start from bptr, go backward and find first endof line
+    * \param bptr end position to go backward
+    * \param begin the beginning position of buffer
+    * \return position of first endof line going backward
+    */
+  inline char *BackFindEndLine(char *bptr, char *begin) {
```
Comment (cjolivier01) [inline src/data/text_parser.h]: If it doesn't use any member variables, please make it static

File: src/data/text_parser.h [hunk 2]
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
...
+        break;
+      if (**begin != '\xBB' && count == 1)
+        break;
+      if (**begin != '\xBF' && count == 2)
+        break;
     }
-    return begin;
+    if (count < 3)
+      *begin = *begin - count;
+    return;
```
Comment (cjolivier01) [inline src/data/text_parser.h]: return not needed here

File: src/data/text_parser.h [hunk 3]
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
   }
 
  protected:
+   /*!
+    * \brief parse data into out
+    * \param begin beginning of buffer
+    * \param end end of buffer
+    */
+  virtual void ParseBlock(char *begin, char *end,
+                           RowBlockContainer<IndexType> *out) = 0;
```
Comment (haojin2) [inline src/data/text_parser.h]: One nit, I guess this line should be aligned with the upper line?

File: src/data/text_parser.h [hunk 4]
```diff
@@ -48,32 +48,52 @@ class TextParserBase : public ParserImpl<IndexType> {
...
+    * \brief read in next several blocks of data
+    * \param data vector of data to be returned
+    * \return true if the data is loaded, false if reach end
+    */
+  inline bool FillData(std::vector<RowBlockContainer<IndexType>> *data);
+   /*!
+    * \brief start from bptr, go backward and find first endof line
+    * \param bptr end position to go backward
+    * \param begin the beginning position of buffer
+    * \return position of first endof line going backward
```
Comment (eric-haibin-lin) [inline src/data/text_parser.h]: clarify that it returns `begin` if not found in (begin, bptr]

**Comment by anirudh2290** [comment]
Thanks a lot @cjolivier01 , @haojin2 and @eric-haibin-lin for reviewing. I have made the necessary changes. Also changed to const char * and moved maxthread initialization outside the parallel region.

**Comment by anirudh2290** [comment]
@tqchen Can you please help review and merge this ?

**Comment by tqchen** [comment]
wait until @cjolivier01 approve and i will merge this in

File: src/data/text_parser.h [hunk 5]
```diff
@@ -48,32 +44,51 @@ class TextParserBase : public ParserImpl<IndexType> {
...
+    */
+  inline bool FillData(std::vector<RowBlockContainer<IndexType>> *data);
+   /*!
+    * \brief start from bptr, go backward and find first endof line
+    * \param bptr end position to go backward
+    * \param begin the beginning position of buffer
+    * \return position of first endof line going backward, returns begin if not found
+    */
+  static inline const char *BackFindEndLine(const char *bptr, const char *begin) {
+     for (; bptr != begin; --bptr) {
```
Comment (larroy) [inline src/data/text_parser.h]: we should assert that bptr >=begin

File: src/data/text_parser.h [hunk 6]
```diff
@@ -48,32 +44,51 @@ class TextParserBase : public ParserImpl<IndexType> {
...
+   * \brief Ignore UTF-8 BOM if present
+   * \param begin reference to begin pointer
+   * \param end reference to end pointer
    */
-  inline char* BackFindEndLine(char *bptr,
-                               char *begin) {
-    for (; bptr != begin; --bptr) {
-      if (*bptr == '\n' || *bptr == '\r') return bptr;
+  static inline void IgnoreUTF8BOM(const char **begin, const char **end) {
+    int count = 0;
```
Comment (larroy) [inline src/data/text_parser.h]: assert that end >= begin?

**Pull request merged**